llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -1820,6 +1821,16 @@ Instruction *InstCombinerImpl::foldOpIntoPhi(Instruction &I, PHINode *PN) {
       continue;
     }
 
+    // If the only use of phi is comparing it with a constant then we can
+    // put this comparison in the incoming BB directly after a ucmp/scmp call
+    // because we know that it will simplify to a single icmp.
+    if (isa<CmpIntrinsic>(InVal) &&
+        match(&I, m_c_ICmp(m_Specific(PN), m_Constant()))) {
```

Comment:
This should also check that the CmpIntrinsic is one-use and m_Constant() should be m_APInt(), because we only fold in that case.